### PR TITLE
[vcpkg] Further JSON error improvements

### PR DESCRIPTION
**Pull request by ras0219**
This PR does a few things:

1. Internally splits the representation (`json.h`) and parser (`jsonreader.h`) to reduce unintended dependencies across the codebase. The same refactor is also applied to Configuration/ConfigurationDeserializer.
2. Introduces levenshtein-distance suggestions for misspelled fields.
3. Minor tweak to the `identifier` error message.

Potential future work includes applying this levenshtein-distance calculation to search results.

As a drive by, this PR documents that comment fields (`$xyz`) are not supported in objects being used as sets with user-defined keys; they are only allowed in "struct-like" objects with a defined keyspace.

### Example 1
```json
{ "name": "Abseil", ... }
```
```
PS C:\src\vcpkg\testing> ..\vcpkg-old.exe install --feature-flags=manifests,registries --dry-run
Errors occurred while parsing C:\src\vcpkg\testing\vcpkg.json
    $.name: mismatched type: expected an identifier
See https://github.com/Microsoft/vcpkg/tree/master/docs/specifications/manifests.md for more information.
```
```
PS C:\src\vcpkg\testing> ..\vcpkg.exe install --feature-flags=manifests,registries --dry-run
Errors occurred while parsing C:\src\vcpkg\testing\vcpkg.json
    $.name (an identifier): must be lowercase alphanumeric+hyphens and not reserved
See https://github.com/Microsoft/vcpkg/tree/master/docs/specifications/manifests.md for more information.
```

### Example 2
```json
{ .., "depednecies": ["foo"], ... }
```
```
PS C:\src\vcpkg\testing> ..\vcpkg-old.exe install --feature-flags=manifests,registries --dry-run
Errors occurred while parsing C:\src\vcpkg\testing\vcpkg.json
    $ (a manifest): unexpected field 'depednecies'
See https://github.com/Microsoft/vcpkg/tree/master/docs/specifications/manifests.md for more information.
```
```
PS C:\src\vcpkg\testing> ..\vcpkg.exe install --feature-flags=manifests,registries --dry-run
Errors occurred while parsing C:\src\vcpkg\testing\vcpkg.json
    $ (a manifest): unexpected field 'depednecies', did you mean 'dependencies'?
See https://github.com/Microsoft/vcpkg/tree/master/docs/specifications/manifests.md for more information.
```

### Example 3
```json
{
  ...,
  "features": {
    "$abc" : true
  }
}
```
```
PS C:\src\vcpkg\testing> ..\vcpkg-old.exe install --feature-flags=manifests,registries --dry-run
Errors occurred while parsing C:\src\vcpkg\testing\vcpkg.json
    $.features (a features field): unexpected field '$abc'
See https://github.com/Microsoft/vcpkg/tree/master/docs/specifications/manifests.md for more information.
```
```
PS C:\src\vcpkg\testing> ..\vcpkg.exe install --feature-flags=manifests,registries --dry-run
Errors occurred while parsing C:\src\vcpkg\testing\vcpkg.json
    $.features (a set of features): unexpected field '$abc': must be lowercase alphanumeric+hyphens and not reserved
See https://github.com/Microsoft/vcpkg/tree/master/docs/specifications/manifests.md for more information.
```

### CODE REVIEW

**Comment by strega-nil** [comment]
Putting myself as a required reviewer.

File: toolsrc/src/vcpkg/base/json.cpp [hunk 1]
```diff
@@ -1292,9 +1323,20 @@ namespace vcpkg::Json
...
+            auto best_value = Strings::byte_edit_distance(f, valid_fields[0]);
+            for (size_t i = 1; i < valid_fields.size(); ++i)
+            {
+                auto v = Strings::byte_edit_distance(f, valid_fields[i]);
+                if (v < best_value)
+                {
+                    best_value = v;
+                    best_index = i;
+                }
+            }
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/base/json.cpp:1338]:
```suggestion
            const auto last = valid_fields.end();
            auto best_it = valid_fields.begin();
            auto best_value = Strings::byte_edit_distance(f, *best_it);
            for (auto it = best_it + 1; it != last; ++it)
            {
                auto v = Strings::byte_edit_distance(f, *it);
                if (v < best_value)
                {
                    best_value = v;
                    best_it = it;
                }
            }
```

I'd prefer you used iterators here.

File: toolsrc/include/vcpkg/configurationparser.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#pragma once
```
Comment (strega-nil) [inline toolsrc/include/vcpkg/configurationparser.h:1]: I do not understand why this is a different file.

File: toolsrc/include/vcpkg/base/view.h [hunk 3]
```diff
@@ -1,9 +1,5 @@
 #pragma once
```
Comment (strega-nil) [inline toolsrc/include/vcpkg/base/view.h:1]: There's definitely no reason for this to be a different file.

File: toolsrc/include/vcpkg/base/strings.h [hunk 4]
```diff
@@ -286,4 +293,6 @@ namespace vcpkg::Strings
 
     // base 32 encoding, following IETC RFC 4648
     std::string b32_encode(std::uint64_t x) noexcept;
+
+    size_t byte_edit_distance(StringView a, StringView b);
```
Comment (strega-nil) [inline toolsrc/include/vcpkg/base/strings.h:297]: Please add comments here on the actual standard followed.

File: toolsrc/include/vcpkg/base/jsonreader.h [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    private:
+        StringLiteral type_name_;
+    };
+
+    struct PathDeserializer final : IDeserializer<fs::path>
+    {
+        virtual StringView type_name() const override { return "a path"; }
+        virtual Optional<fs::path> visit_string(Reader&, StringView sv) override { return fs::u8path(sv); }
+
+        static PathDeserializer instance;
```
Comment (strega-nil) [inline toolsrc/include/vcpkg/base/jsonreader.h:224]:
```suggestion
        static const PathDeserializer instance;
```

Could we do this, and also make all the `visit` functions `const`? There's no reason for them to be non-`const`.

(same for the rest of the instances)

File: toolsrc/src/vcpkg/base/json.cpp [hunk 6]
```diff
@@ -1,5 +1,6 @@
 #include <vcpkg/base/files.h>
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/base/json.cpp:1]: Could we add a cpp file here? It's weird to me that we've split the headers but not the cpp files.

File: toolsrc/src/vcpkg/base/strings.cpp [hunk 7]
```diff
@@ -258,6 +258,43 @@ bool Strings::contains(StringView haystack, StringView needle)
...
+    {
+        char diag = d[0];
+        d[0] = static_cast<char>(std::min(d[0] + 1, j + (pa[0] != pb[j])));
+        for (int i = 1; i < sa; ++i)
+        {
+            auto subst_or_add = std::min(d[i - 1] + 1, diag + (pa[i] != pb[j]));
+            d[i] = static_cast<char>(std::min((diag = d[i]) + 1, subst_or_add));
+        }
+    }
+    return d[sa - 1];
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/base/strings.cpp:295]: This code is extremely ugly, and not obviously UB-correct. Can we rewrite it to be more obviously correct and C++-y?

File: toolsrc/src/vcpkg/configuration.cpp [hunk 8]
```diff
@@ -26,12 +28,10 @@ namespace vcpkg
             }
         }
 
+        static Json::ArrayDeserializer<RegistryDeserializer> arr_reg_d{"an array of registries"};
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/configuration.cpp:31]:
```suggestion
        static Json::ArrayDeserializer<RegistryDeserializer> array_of_registries{"an array of registries"};
```

File: toolsrc/src/vcpkg/registries.cpp [hunk 9]
```diff
@@ -50,22 +53,22 @@ namespace vcpkg
     Optional<std::unique_ptr<RegistryImpl>> RegistryImplDeserializer::visit_object(Json::Reader& r,
                                                                                    const Json::Object& obj)
     {
+        static Json::StringDeserializer kind_des{"a registry implementation kind"};
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/registries.cpp:56]:
```suggestion
        static Json::StringDeserializer kind_deserializer{"a registry implementation kind"};
```

File: toolsrc/src/vcpkg/registries.cpp [hunk 10]
```diff
@@ -98,13 +102,10 @@ namespace vcpkg
             return nullopt;
         }
 
+        static Json::ArrayDeserializer<Json::PackageNameDeserializer> arr_reg_d{"an array of package names"};
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/registries.cpp:105]:
```suggestion
        static Json::ArrayDeserializer<Json::PackageNameDeserializer> package_names_deserializer{"an array of package names"};
```

File: toolsrc/src/vcpkg/sourceparagraph.cpp [hunk 11]
```diff
@@ -371,11 +372,14 @@ namespace vcpkg
             }
             else
             {
-                Debug::print("Failed to parse platform expression: ", opt.error(), "\n");
-                return nullopt;
+                r.add_generic_error(type_name(), opt.error());
+                return PlatformExpression::Expr::Empty();
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/sourceparagraph.cpp:376]:
```suggestion
                return nullopt;
```

This is an error state, there's no reason to return a default constructed state.

File: toolsrc/src/vcpkg/sourceparagraph.cpp [hunk 12]
```diff
@@ -688,45 +707,41 @@ namespace vcpkg
...
+            r.optional_object_field(obj, SUPPORTS, spgh->supports_expression, PlatformExprDeserializer::instance);
 
             r.optional_object_field(
-                obj,
-                FEATURES,
-                control_file->feature_paragraphs,
-                Json::ArrayDeserializer<FeatureDeserializer>{"an array of feature definitions", Json::AllowEmpty::Yes});
+                obj, DEFAULT_FEATURES, spgh->default_features, Json::IdentifierArrayDeserializer::instance);
+
+            static Json::ArrayDeserializer<FeatureDeserializer> arr_feature_d{"an array of feature definitions"};
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/sourceparagraph.cpp:735]:
```suggestion
            static Json::ArrayDeserializer<FeatureDeserializer> features_deserializer{"an array of feature definitions"};
```

**Review by strega-nil — CHANGES_REQUESTED** [review]

File: toolsrc/include/vcpkg/configurationparser.h (hunk 2 above)
Comment (ras0219-msft) [reply toolsrc/include/vcpkg/configurationparser.h:1]: Consumers of Configuration objects do not care how they were produced; this is splitting the production (deserializers) from consumption. Similarly, the only code that uses `RegistryImplDeserializer` also uses `ConfigurationDeserializer`, so it makes sense to put them in the same file.

File: toolsrc/src/vcpkg/base/strings.cpp (hunk 7 above)
Comment (ras0219-msft) [reply toolsrc/src/vcpkg/base/strings.cpp:295]: I think that this style is optimal for the algorithm, however I'll try to add more comments to help the reader optically prove correctness.

File: toolsrc/src/vcpkg/sourceparagraph.cpp (hunk 11 above)
Comment (ras0219-msft) [reply toolsrc/src/vcpkg/sourceparagraph.cpp:376]: The reason to return a default constructed state is to indicate to the _parent_ code that the error has already been handled. Otherwise, you get two errors emitted for the same "problem": a specific error (emitted on the line above) and a generic error (emitted in the parent code).

With the previous JSON refactor, Deserializers should either:
1. Emit `nullopt` when they have no clue what's going on and don't have more specific error information to provide
2. Emit a more specific error with context information and return some default-constructed value.

File: toolsrc/include/vcpkg/configurationparser.h (hunk 2 above)
Comment (strega-nil) [reply toolsrc/include/vcpkg/configurationparser.h:1]: hm, alright. However, you should name the file `configuration-parser.h`; I hate the existing practice of "ccaseeverythingisonelowercaseword"

File: toolsrc/src/vcpkg/base/strings.cpp (hunk 7 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/strings.cpp:295]: You definitely at least want to switch to `size_t` from `int` for the indexing, since sext is slow.

File: toolsrc/src/vcpkg/sourceparagraph.cpp (hunk 11 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/sourceparagraph.cpp:376]: hmm, alright.

File: toolsrc/include/vcpkg/configurationparser.h (hunk 2 above)
Comment (ras0219) [reply toolsrc/include/vcpkg/configurationparser.h:1]: This matches our current practice:
- `portfileprovider.h`
- `sourceparagraph.h`
- `paragraphparser.h`
- `vcpkgcmdarguments.h`
- `vcpkgpaths.h`

However I've renamed it to `configurationdeserializer.h` to better match the principle type

File: toolsrc/include/vcpkg/base/jsonreader.h (hunk 5 above)
Comment (ras0219) [reply toolsrc/include/vcpkg/base/jsonreader.h:224]: This is true for now, however I didn't want to assume that all deserializers will forever not need mutable members.

Regardless, I think if we want to make this change it can be made in a separate PR.

File: toolsrc/src/vcpkg/sourceparagraph.cpp (hunk 11 above)
Comment (ras0219) [reply toolsrc/src/vcpkg/sourceparagraph.cpp:376]: That's an equivalent option, but I don't think it significantly improves the clarity of the code. It just pushes calling `r.add_error()` into the parent code, yet requires the formatting to be done in the child code. Since the context is omnipresent throughout the call stack, it's just as easy to simply call it directly and return a default value.

The further alternative would be to go all the way to some `ExpectedT<U, std::vector<string>>`, however this mandates a bunch of individual array allocations which makes me 😢.

File: toolsrc/include/vcpkg/base/fwd/json.h [hunk 13]
```diff
@@ -12,8 +12,6 @@ namespace vcpkg::Json
     struct Object;
     struct Array;
 
-    struct ReaderError;
-    struct BasicReaderError;
     struct Reader;
```
[the remark this answers is not in the document]
Comment (ras0219) [reply toolsrc/include/vcpkg/base/fwd/json.h]: Moved to just `jsonreader.h`

File: toolsrc/include/vcpkg/base/fwd/json.h [hunk 14]
```diff
@@ -23,17 +21,21 @@ namespace vcpkg::Json
         using type = Type;
```
[the remark this answers is not in the document]
Comment (ras0219) [reply toolsrc/include/vcpkg/base/fwd/json.h]: Moved to just `jsonreader.h`

File: toolsrc/include/vcpkg/base/view.h (hunk 3 above)
Comment (ras0219) [reply toolsrc/include/vcpkg/base/view.h:1]: This already exists and would require touching significantly more files to change; if we choose to remove it, we should do it in a different PR.

File: toolsrc/src/vcpkg/base/json.cpp (hunk 6 above)
Comment (ras0219) [reply toolsrc/src/vcpkg/base/json.cpp:1]: I don't feel that moving 130 lines of code into a new file is worth the additional 4+ cpu-seconds of compile time and merge conflicts with #13421.

File: toolsrc/include/vcpkg/base/jsonreader.h (hunk 5 above)
Comment (strega-nil) [reply toolsrc/include/vcpkg/base/jsonreader.h:256]: Global mutable state is an extreme code smell, though... even though it's fine here, I don't like that we have it at all.

File: toolsrc/include/vcpkg/base/jsonreader.h [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+    template<class Type>
+    struct IDeserializer
+    {
+        using type = Type;
+        virtual StringView type_name() const = 0;
+
+    private:
+        friend struct Reader;
+        Optional<Type> visit(Reader&, const Value&);
+        Optional<Type> visit(Reader&, const Object&);
```
Comment (strega-nil) [inline toolsrc/include/vcpkg/base/jsonreader.h:24]: why was this moved over?

File: toolsrc/src/vcpkg/base/json.cpp (hunk 6 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/json.cpp:1]: It really adds 4 seconds? that's shocking...

File: toolsrc/include/vcpkg/configurationdeserializer.h [hunk 16]
```diff
@@ -0,0 +1,66 @@
+#pragma once
```
Comment (strega-nil) [inline toolsrc/include/vcpkg/configurationdeserializer.h:1]: Please rename this file to `configuration-deserializer.h`

File: toolsrc/src/vcpkg/sourceparagraph.cpp [hunk 17]
```diff
@@ -688,45 +707,42 @@ namespace vcpkg
                 }
             }
 
+            static Json::StringDeserializer version_d{"a version"};
+            static Json::StringDeserializer url_d{"a url"};
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/sourceparagraph.cpp:711]: Please write out the full words

```suggestion
            static Json::StringDeserializer version_deserializer{"a version"};
            static Json::StringDeserializer url_deserializer{"a url"};
```

File: toolsrc/src/vcpkg/sourceparagraph.cpp [hunk 18]
```diff
@@ -688,45 +707,42 @@ namespace vcpkg
...
+            r.optional_object_field(obj, DESCRIPTION, spgh->description, Json::ParagraphDeserializer::instance);
+            r.optional_object_field(obj, HOMEPAGE, spgh->homepage, url_d);
+            r.optional_object_field(obj, DOCUMENTATION, spgh->documentation, url_d);
+            r.optional_object_field(obj, LICENSE, spgh->license, LicenseExpressionDeserializer::instance);
+            r.optional_object_field(obj, DEPENDENCIES, spgh->dependencies, DependencyArrayDeserializer::instance);
 
             if (obj.contains(DEV_DEPENDENCIES))
             {
-                System::print2(System::Color::error, "dev_dependencies are not yet supported");
+                System::print2(System::Color::error, DEV_DEPENDENCIES, " are not yet supported");
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/sourceparagraph.cpp:726]: :+1:

**Review by strega-nil — COMMENTED** [review]
Just a few things left :)

File: toolsrc/include/vcpkg/base/jsonreader.h (hunk 15 above)
Comment (ras0219) [reply toolsrc/include/vcpkg/base/jsonreader.h:24]: Dependency reduction -- the number of places that require knowledge of the `Json::Reader` is much smaller than the number of places that require `Json::Value`. In fact, there are currently _no_ headers that require it (search for `#include <vcpkg/base/jsonreader.h>`).

File: toolsrc/include/vcpkg/configurationdeserializer.h (hunk 16 above)
Comment (ras0219) [reply toolsrc/include/vcpkg/configurationdeserializer.h:1]: No, see my response to your comment above. The current style of the codebase is to lowercase type names and this PR is not about changing the current codebase style.

File: toolsrc/src/vcpkg/base/json.cpp (hunk 6 above)
Comment (ras0219) [reply toolsrc/src/vcpkg/base/json.cpp:1]: Remember that every cpp file includes the _entire STL_ via the PCH.

**Review by strega-nil — APPROVED** [review]

**Review by strega-nil — APPROVED** [review]
